File: ios/Astro/Astro/Models/AstroLocation.swift
//
//  WorldCity.swift
//  Astro
//
//  Created by Aldana, Sal on 1/19/21.
//

import Foundation
import GRDB
import MapKit

//MARK: - Locations Table -
class AstroLocation: Record, Identifiable {
    var id: Int
    var city: String
    var country: String
    var timezone: String
    var coordinates: CLLocationCoordinate2D
    var location: CLLocation {
        return CLLocation(latitude: coordinates.latitude, longitude: coordinates.longitude)
    }
    
    override class var databaseTableName: String {
        return "locations"
    }
    
    enum Columns: String, ColumnExpression {
        case id, country, city, latitude, longitude, timezone
    }
    
    required init(row: Row) {
        self.id = row[Columns.id]
        self.city = row[Columns.city]
        self.country = row[Columns.country]
        self.timezone = row[Columns.timezone]
        let lat: Double = row[Columns.latitude]
        let lng: Double = row[Columns.longitude]
        self.coordinates = CLLocationCoordinate2D(latitude: lat, longitude: lng)
        super.init(row: row)
    }
    
    init(id: Int, city: String, country: String, timeZone: String, lat: Double, lng: Double) {
        self.id = id
        self.city = city
        self.country = country
        self.timezone = timeZone
        self.coordinates = CLLocationCoordinate2D(latitude: lat, longitude: lng)
        super.init()
    }
    
    static var defaultLocation: AstroLocation {
        return AstroLocation(id: 297,
                             city: "San Francisco",
                             country: "United States",
                             timeZone: "America/Los_Angeles",
                             lat: 37.7749295,
                             lng: -122.4194183)
    }
    
    static var defaultSpan: MKCoordinateSpan {
        return MKCoordinateSpan(latitudeDelta: 12,
                                longitudeDelta: 12)
    }
    
    static var defaultRegion: MKCoordinateRegion {
        return MKCoordinateRegion(center: AstroLocation.defaultLocation.coordinates,
                                  span: AstroLocation.defaultSpan)
    }
}

File: android/Astro/app/src/main/java/net/level3studios/astro/models/AstroLocation.kt
package net.level3studios.astro.models

import android.database.Cursor
import android.location.Location
import com.google.android.gms.maps.model.LatLng

class AstroLocation() {
    var id: Int = 0
    var city: String = ""
    var country: String = ""
    var timeZone: String = ""
    var coordinates: LatLng = LatLng(0.0,0.0)

    companion object {
        fun defaultLocation(): AstroLocation {
            val sanFran = AstroLocation()
            sanFran.id = 297
            sanFran.city = "San Francisco"
            sanFran.country = "United States"
            sanFran.timeZone = "America/Los_Angeles"
            sanFran.coordinates = LatLng(37.7749295, -122.4194183)
            return sanFran
        }
    }

    enum class Columns(var label: String) {
        ID("id"),
        CITY("city"),
        COUNTRY("country"),
        LATITUDE("latitude"),
        LONGITUDE("longitude"),
        TIMEZONE("timezone")
    }

    constructor(row: Cursor) : this() {
        this.id = row.getInt(row.getColumnIndex(Columns.ID.label))
        this.city = row.getString(row.getColumnIndex(Columns.CITY.label))
        this.country = row.getString(row.getColumnIndex(Columns.COUNTRY.label))
        this.timeZone = row.getString(row.getColumnIndex(Columns.TIMEZONE.label))
        val lat = row.getDouble(row.getColumnIndex(Columns.LATITUDE.label))
        val lng = row.getDouble(row.getColumnIndex(Columns.LONGITUDE.label))
        this.coordinates = LatLng(lat, lng)
    }
}